src/main/scala/com/github/sstone/amqp/ChannelOwner.scala

```diff
@@ -176,50 +176,65 @@ class ChannelOwner(init: Seq[Request] = Seq.empty[Request], channelParams: Optio
   }
 
   def onChannel(channel: Channel, forwarder: ActorRef): Unit = {
-    channelParams.map(p => channel.basicQos(p.qos, p.global))
+    channelParams.foreach(p => channel.basicQos(p.qos, p.global))
   }
 
   def receive = disconnected
 
-  def disconnected: Receive = LoggingReceive {
-    case channel: Channel => {
-      val forwarder = context.actorOf(Props(new Forwarder(channel)), name = "forwarder-" + UUID.randomUUID.toString)
-      forwarder ! AddShutdownListener(self)
-      forwarder ! AddReturnListener(self)
-      onChannel(channel, forwarder)
-      requestLog.map(r => self forward r)
-      log.info(s"got channel $channel")
-      statusListeners.map(a => a ! Connected)
-      context.become(connected(channel, forwarder))
+  def newChannel(channel: Channel, previousForwarder: Option[ActorRef]) = {
+    log.info(s"got channel $channel")
+    previousForwarder.foreach { forwarder =>
+      context.stop(forwarder)
+      statusListeners.foreach(_ ! Disconnected)
     }
-    case Record(request: Request) => {
-      requestLog :+= request
+    val forwarder = context.actorOf(Props(new Forwarder(channel)), name = "forwarder-" + UUID.randomUUID.toString)
+    forwarder ! AddShutdownListener(self)
+    forwarder ! AddReturnListener(self)
+    Try(onChannel(channel, forwarder)) match {
+      case Success(_) =>
+        requestLog.foreach(r => self forward r)
+        if (previousForwarder.isEmpty)
```

Comment:
Shouldn't we be sending the Connected message despite being connected before?

Reply:
Yes, of course! Fixed in 036f91e.